omnilib/control.lua

```diff
@@ -1,408 +1,343 @@
-						if grade and catRec[category][prefix].level and grade > catRec[category][prefix].level then catRec[category][prefix].level = grade end
-						if not inTable(suffix,catRec[category][prefix].suffixes) then table.insert(catRec[category][prefix].suffixes,suffix) end
-					end
-				end
-			end
-			local compressionSuffix = {"","-compression"}
-			for cname,c in pairs(catRec) do
-				if global and global.omni and global.omni.cat and global.omni.cat[cname] then
-					if settings.global["omnilib-autoupdate"].value == true then
-						for _,kind in pairs(global.omni.cat[cname].buildings) do
-							for _,surface in pairs(game.surfaces) do
-								for _,entity in pairs(surface.find_entities_filtered{force=event.research.force, name=kind}) do
-									log(kind)
-									local rec = entity.get_recipe()
-									if rec then
-										local ingredients = {inventory={},fluidbox={}}
-										if entity.is_crafting() then
-											for _,ing in pairs(rec.ingredients) do
-												if ing.type=="fluid" then
-													table.insert(ingredients.fluidbox,ing)
-												else
-													table.insert(ingredients.inventory,ing)
-												end
-											end
-											--[[table.insert(recipeDeficiet,{
-												entity=entity,
-												ingredients = ingredients
-											})]]
-										end
-										for prefix,recipe in pairs(c) do
-											if recipe.level > 1 then
-												for _,suffix in pairs(recipe.suffixes) do
-													for _,comp in pairs(compressionSuffix) do
-														if rec.name == prefix..ord[recipe.level-1]..comp..suffix then
-															for _, sufTarget in pairs(recipe.suffixes) do
-																if tech.force.recipes[prefix..ord[recipe.level]..comp..sufTarget] and (suffix==sufTarget or not tech.force.recipes[prefix..ord[recipe.level]..comp..suffix]) then
-																	entity.set_recipe(prefix..ord[recipe.level]..comp..sufTarget)
-																	break
-																end
-															end
-														end
-													end
-												end
-											end
-										end
-									end
-								end
-							end
-						end
-					end
-					for prefix,recipe in pairs(c) do
-						if recipe.level then
-							for i=1,recipe.level-1 do
-								for _,suffix in pairs(recipe.suffixes) do
-									for _,comp in pairs(compressionSuffix) do
-										if tech.force.recipes[prefix..ord[i]..comp..suffix] then
-											tech.force.recipes[prefix..ord[i]..comp..suffix].enabled = false
-										end
-									end
-								end
-							end
-						end
-					end
-				end
+			global.omni.need_update = true
+		elseif not global.omni.need_update then -- If we need_update we'll be iterating everything anyway.
+			update_tech(tech, true)
+			if tech.name:find("^omnitech") and not tech.name:find("^omnipressed") then
+				global.omni.update_buildings = true
 			end
 		end
 	end
 end)
 
+script.on_event(defines.events.on_technology_effects_reset, function(event)
```

Comment:
I think this has a possibility of creating an infinite loop, if something locks a tech we'd normally unlock and then resets tech effects, feedback requested.